[Heartbeat] Add functional-ish tests in golang

andrewvc opened the pull request:
Fixes #31999

This PR turned out to be a bit of a monster. The main thing it does is add testing for functional 'scenarios', that is testing full heartbeat configs and the events they output going through all heartbeat specific code using the libbeat factory interface. This lets us do something sort of close to functional testing, where instead of using the heartbeat binary we actually use this factory interface. This gets us pretty close with a much simpler testing experience that's easier to debug and faster to run. It probably won't miss any error conditions from generating the full binary.

This also cleans up a bunch of linter issues in modified files, which is a bit distracting, things like missing error return values, shadowed variable names etc.

Finally, the new scenario test this adds, which checks that all events always contain a check group, revealed two bugs:
1. Synthetics test runs create an empty event at the start of each run (with no check group)
2. Any synth events prior to a journey start event would be missing a check group

This PR fixes this by:
1. Moving the check group generation code for synthexec up to the stream enricher, rather than the journey enricher.
2. Creating the check group UUIDs after journeys end, rather than waiting for them to start. The synthetics node agent can produce output before any journey starts



## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-32542/detail/PR-32542/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-32542/detail/PR-32542/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-32542/detail/PR-32542/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-32542/detail/PR-32542/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_32542.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-07-28T20:50:20.512Z&rangeTo=2022-07-28T21:10:20.512Z&transactionName=Beats/beats/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=b0cdb8298b688c131d5b2b903420f7a9&transactionId=b6098ed07a7fe24d)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-07-28T21:00:20.512+0000


* Duration: 14 min 46 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 3  |
| Skipped      | 0 |
| Total        | 3   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![4](https://img.shields.io/badge/4%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-32542/detail/PR-32542/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `heartbeat-checks - make -C heartbeat check;make -C heartbeat update;make -C x-pack/heartbeat check`
<ul>
<li>Took 0 min 9 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-32542/runs/1/steps/1301/log/?start=0">here</a></li>
<li>Description: <code>make -C heartbeat check;make -C heartbeat update;make -C x-pack/heartbeat check;make -C x-pack/heartbeat update;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `x-pack/heartbeat-checks - make -C x-pack/heartbeat check;make -C x-pack/heartbeat update;make -C h`
<ul>
<li>Took 2 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-32542/runs/1/steps/1295/log/?start=0">here</a></li>
<li>Description: <code>make -C x-pack/heartbeat check;make -C x-pack/heartbeat update;make -C heartbeat check;make -C heartbeat update;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `make check-default`
<ul>
<li>Took 2 min 34 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-32542/runs/1/steps/872/log/?start=0">here</a></li>
<li>Description: <code>make check-default</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-32542/runs/1/steps/1312/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 2'</code></l1>
</ul>
  
  </p>
  </details>



## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

andrewvc commented on heartbeat/beater/heartbeat.go:
```diff
@@ -67,23 +69,36 @@ func New(b *beat.Beat, rawConfig *conf.C) (beat.Beater, error) {
 	jobConfig := parsedConfig.Jobs
 
-	scheduler := scheduler.Create(limit, hbregistry.SchedulerRegistry, location, jobConfig, parsedConfig.RunOnce)
+	sched := scheduler.Create(limit, hbregistry.SchedulerRegistry, location, jobConfig, parsedConfig.RunOnce)
```
renamed this variable to prevent shadowing the `scheduler` package

andrewvc commented on heartbeat/beater/heartbeat.go:
```diff
@@ -67,23 +69,36 @@ func New(b *beat.Beat, rawConfig *conf.C) (beat.Beater, error) {
-	scheduler := scheduler.Create(limit, hbregistry.SchedulerRegistry, location, jobConfig, parsedConfig.RunOnce)
+	sched := scheduler.Create(limit, hbregistry.SchedulerRegistry, location, jobConfig, parsedConfig.RunOnce)
+
+	pipelineClientFactory := func(p beat.Pipeline) (pipeline.ISyncClient, error) {
```
Moved the creation of the pipeline client into heartbeat

andrewvc commented on heartbeat/beater/heartbeat.go:
```diff
@@ -67,23 +69,36 @@ func New(b *beat.Beat, rawConfig *conf.C) (beat.Beater, error) {
 // Run executes the beat.
 func (bt *Heartbeat) Run(b *beat.Beat) error {
-	logp.Info("heartbeat is running! Hit CTRL-C to stop it.")
+	logp.L().Info("heartbeat is running! Hit CTRL-C to stop it.")
```
Fixed up the logger uses to avoid deprecation warnings

andrewvc commented on heartbeat/beater/heartbeat.go:
```diff
@@ -108,7 +123,7 @@ func (bt *Heartbeat) Run(b *beat.Beat) error {
 		bt.monitorReloader = cfgfile.NewReloader(b.Publisher, bt.config.ConfigMonitors)
 		defer bt.monitorReloader.Stop()
 
-		err := bt.RunReloadableMonitors(b)
```
This variable was unused

andrewvc commented on heartbeat/beater/heartbeat.go:
```diff
@@ -166,14 +181,14 @@ func (bt *Heartbeat) RunStaticMonitors(b *beat.Beat) (stop func(), err error) {
 
 // RunReloadableMonitors runs the `heartbeat.config.monitors` portion of the yaml config if present.
-func (bt *Heartbeat) RunReloadableMonitors(b *beat.Beat) (err error) {
+func (bt *Heartbeat) RunReloadableMonitors() (err error) {
```
removed unused variable

andrewvc commented on heartbeat/beater/heartbeat.go:
```diff
@@ -166,14 +181,14 @@ func (bt *Heartbeat) RunStaticMonitors(b *beat.Beat) (stop func(), err error) {
 func (bt *Heartbeat) RunCentralMgmtMonitors(b *beat.Beat) {
-	monitors := cfgfile.NewRunnerList(management.DebugK, bt.dynamicFactory, b.Publisher)
-	reload.Register.MustRegisterList(b.Info.Beat+".monitors", monitors)
+	mons := cfgfile.NewRunnerList(management.DebugK, bt.dynamicFactory, b.Publisher)
```
fix variable name shadowing package `monitors`

andrewvc commented on heartbeat/beater/heartbeat.go:
```diff
@@ -187,7 +202,7 @@ func (bt *Heartbeat) RunReloadableMonitors(b *beat.Beat) (err error) {
 // makeAutodiscover creates an autodiscover object ready to be started.
 func (bt *Heartbeat) makeAutodiscover(b *beat.Beat) (*autodiscover.Autodiscover, error) {
-	autodiscover, err := autodiscover.NewAutodiscover(
+	autodisco, err := autodiscover.NewAutodiscover(
```
fix variable name shadowing package

andrewvc commented on heartbeat/monitors/factory.go:
```diff
@@ -40,15 +41,17 @@ import (
-	mtx        *sync.Mutex
-	pluginsReg *plugin.PluginsReg
-	logger     *logp.Logger
-	runOnce    bool
```
runOnce is no longer needed since we determine the client type in heartbeat.go and pass in the right client creation function

andrewvc commented on heartbeat/monitors/factory.go:
```diff
@@ -40,15 +41,17 @@ import (
+	mtx                   *sync.Mutex
+	pluginsReg            *plugin.PluginsReg
+	logger                *logp.Logger
+	pipelineClientFactory PipelineClientFactory
```
We now take this factory that returns either a sync or regular client

andrewvc commented on heartbeat/monitors/factory_test.go:
```diff
@@ -164,14 +161,13 @@ func TestDuplicateMonitorIDs(t *testing.T) {
 	serverMonConf := mockPluginConf(t, "custom", "custom", "@every 1ms", "http://example.net")
 	badConf := mockBadPluginConf(t, "custom")
 	reg, built, closed := mockPluginsReg()
-	pipelineConnector := &MockPipelineConnector{}
```
PipelineConnector == Pipeline, so this is easier to read IMHO

andrewvc commented on heartbeat/monitors/mocks.go:
```diff
@@ -22,42 +22,98 @@ import (
 	"regexp"
```
This file is now not a `_test` file because it's used as a helper in other tasks elsewhere. I wish `_test` files could be used by `_test` files from other packages, but alas it's not possible (for some good reasons)

andrewvc commented on heartbeat/monitors/monitor.go:
```diff
@@ -227,31 +219,7 @@ func (m *Monitor) Start() {
-				wait:    func() {},
-			})
-		}
+		t.Start(m.pubClient)
```
Much cleaner now with the simpler types!

andrewvc commented on x-pack/heartbeat/monitors/browser/source/local.go:
```diff
@@ -56,7 +56,7 @@ func (l *LocalSource) Fetch() (err error) {
 		return nil
 	}
-	l.workingPath, err = ioutil.TempDir("/tmp", "elastic-synthetics-")
+	l.workingPath, err = ioutil.TempDir(os.TempDir(), "elastic-synthetics-")
```
This makes running tests natively on windows possible, and is more correct anyway

andrewvc commented on x-pack/heartbeat/monitors/browser/source/zipurl_test.go:
```diff
@@ -44,7 +44,7 @@ func TestSimpleCases(t *testing.T) {
 				"folder":           "/",
 				"retries":          3,
-				"target_directory": "/tmp/synthetics/blah",
+				"target_directory": filepath.Join(os.TempDir(), "synthetics", "blah"),
```
This makes running tests natively on windows possible, and is more correct anyway, we should always use `filepath` to join dirs where possible.

andrewvc commented on x-pack/heartbeat/monitors/browser/synthexec/enrich.go:
```diff
@@ -23,19 +23,22 @@ import (
 
 func newStreamEnricher(sFields stdfields.StdMonitorFields) *streamEnricher {
-	return &streamEnricher{sFields: sFields}
+	return &streamEnricher{sFields: sFields, checkGroup: makeUuid()}
```
As described in the main issue description, the UUID is now a member of the stream, and changes when each journey ends, removing periods before journey starts where no UUID is present.

andrewvc commented on x-pack/heartbeat/monitors/browser/synthexec/enrich_test.go:
```diff
@@ -208,24 +207,23 @@ func TestEnrichConsoleSynthEvents(t *testing.T) {
 func TestEnrichSynthEvent(t *testing.T) {
 	tests := []struct {
 		name    string
-		je      *journeyEnricher
```
This was really just unused

elasticmachine commented:
Pinging @elastic/uptime (Team:Uptime)

andrewvc commented:
/test

andrewvc commented:
@mergifyio update

lucasfcosta reviewed (commented):
Overall, I didn't have any blocking comments. I tried really hard to see how `mpx` could become `nil`,  but I couldn't see any way for it to become `nil`.

Also, it's quite annoying Jenkins is not showing you any of the logs 😢  I think we may have to use `fmt` here.

I've just started a `repeat 10000` for `dlv` with a conditional breakpoint to see if I can catch this failure locally in a bit and debug.

andrewvc commented on x-pack/heartbeat/scenarios/framework.go:
```diff
@@ -0,0 +1,181 @@
+	}
+
+	t.Run(s.Name, func(t *testing.T) {
+		//t.Parallel()
```
Hmmm, unfortunately I don't think this was it, will re-enable

andrewvc commented on x-pack/heartbeat/monitors/browser/synthexec/synthexec.go:
```diff
@@ -174,6 +177,10 @@ func runCmd(
 	}
 	wg.Add(1)
 	go func() {
+		logp.L().Info("soutpipe %v", stdoutPipe)
+		logp.L().Info("stdoutToSynthEvent %v", stdoutToSynthEvent)
+		logp.L().Info("mpx %v", mpx)
+		logp.L().Info("mpx %v", mpx.writeSynthEvent)
```
Yeah, this was just for debugging that issue, will remove

andrewvc commented on heartbeat/hbtestllext/isdefs.go:
```diff
@@ -41,3 +41,11 @@ var IsInt64 = isdef.Is("positiveInt64", func(path llpath.Path, v interface{}) *l
+var IsUint16 = isdef.Is("positiveUInt16", func(path llpath.Path, v interface{}) *llresult.Results {
+	_, ok := v.(uint16)
+	if !ok {
+		return llresult.SimpleResult(path, false, "expected an int64")
```
🤦 good catch!

andrewvc commented on heartbeat/beater/heartbeat.go:
```diff
@@ -187,7 +202,7 @@ func (bt *Heartbeat) RunReloadableMonitors(b *beat.Beat) (err error) {
 // makeAutodiscover creates an autodiscover object ready to be started.
 func (bt *Heartbeat) makeAutodiscover(b *beat.Beat) (*autodiscover.Autodiscover, error) {
-	autodiscover, err := autodiscover.NewAutodiscover(
+	autodisco, err := autodiscover.NewAutodiscover(
```
yes, great idea

andrewvc commented on x-pack/heartbeat/scenarios/framework.go:
```diff
@@ -0,0 +1,181 @@
+		Hostname:        "localhost",
+		ID:              id,
+		EphemeralID:     eid,
+		FirstStart:      time.Now(),
+		StartTime:       time.Now(),
```
TBH, this info isn't actually used anywhere AFAIK, usually in tests all the fields are just left undefined. I also don't think any delta here would be meaningful, so in short, I'm 99% sure we're good.

andrewvc commented on x-pack/heartbeat/scenarios/framework.go:
```diff
@@ -0,0 +1,181 @@
+	return mtr, err
+}
+
+func makeTestFactory() (factory *monitors.RunnerFactory, sched *scheduler.Scheduler, close func()) {
```
Oh, I guess this is confusing, it isn't a factory for making tests, it's a function that makes factories. Renamed to `setupFactoryAndSched`

andrewvc commented:
I think I've addressed all your comments. I also added a fix in synthexec for a situation where the command can't be found. I think this may have some association with the weirdness we saw with browser monitors, but I also fixed the code to not try to run browser monitors on CI. We never should have been trying since the synthetics agent isn't present on CI.

I'm fine leaving these tests effectively disabled for now until such time as we fix that, at which point we can debug further. Esp. since locally they still run.

lucasfcosta commented on x-pack/heartbeat/monitors/browser/browser.go:
```diff
@@ -27,6 +28,8 @@ func create(name string, cfg *config.C) (p plugin.Plugin, err error) {
 	// this flag. When we're ready to support the many possible configurations of systems outside the docker environment
 	// we can remove this check.
+	ev := os.Environ()
+	logp.L().Info("EV %v", ev)
```
Do we want to keep this log line here? It feels to me we shouldn't log all the env. vars to `stdout` as that could leak information to logs.

andrewvc replied:
No, good catch! I thought I'd pulled all of those

lucasfcosta commented on x-pack/heartbeat/monitors/browser/synthexec/synthexec.go:
```diff
@@ -263,7 +272,7 @@ func runCmd(
 	go func() {
 		err := <-cmdDone
-		jsonWriter.Close()
+		_ = jsonWriter.Close()
```
Any particular reason for us to do this assignment? Maybe it's my lack of familiarity with Go, but it doesn't seem it makes a difference in this case.

andrewvc replied:
It makes intellij's linter happier. `_ = ` means discard this. It's the only assignment that never uses `:=` because nothing is actually assigned. See https://www.geeksforgeeks.org/what-is-blank-identifierunderscore-in-golang/

TBH, I'm not sure I should be changing this vs. changing my intellij settings but it's certainly not harmful.

andrewvc replied:
to be clear, the linter complains about ignored error values, since go doesn't have exceptions

lucasfcosta commented on x-pack/heartbeat/monitors/browser/synthexec/synthexec.go:
```diff
@@ -198,7 +205,9 @@ func runCmd(
 	wg.Add(1)
 	go func() {
-		defer jsonReader.Close()
+		defer func() {
+			_ = jsonReader.Close()
+		}()
```
Can't we defer as we were doing previously?

I tried to do some research to understand the difference between this and the previous `defer` pattern and I couldn't find anything that would explain it.

[I found this response on SO interesting](https://stackoverflow.com/questions/32541870/multiple-defers-vs-deferred-anonymous-function) but it actually advocates for the opposite pattern: directly deferring (although I can see the advantage of a lambda if we're doing any error handling for the deferred function.

[Maybe you're doing it to evaluate `jsonReader`](https://yourbasic.org/golang/defer/) only when the function exits? I couldn't see why that would be helpful in this case though since we only assign to it with `os.Pipe()` at the beginning of the function?

If this is indeed intentional perhaps it would be worth leaving a comment as to why we do it.

andrewvc replied:
Same story WRT the linter and errors

lucasfcosta reviewed (approved):
LGTM! Also cloned the repo and ran the functional tests locally to make sure they were running smoothly.

Although I have approved this, please do have a look at my comment about logging the `os.Environ`. I do think that's important to address and it would be great if you could check that comment and see if you think it's worth doing indeed.

All other comments are not blocking.

andrewvc commented:
@Mergifyio backport 8.4

vigneshshanmugam commented on x-pack/heartbeat/monitors/browser/synthexec/enrich.go:
```diff
@@ -206,6 +205,9 @@ func (je *journeyEnricher) createSummary(event *beat.Event) error {
 	}
 
+	// create a new check group for the next journey
+	je.streamEnricher.checkGroup = makeUuid()
```
Question: Would check-group be unique if we run zip urls which could contain multiple journeys since we moved the generation to the stream enricher?

andrewvc replied:
This code should still create multiple check groups for zip URLs. I don't think we want to assert this at the new functional tests though, since my hope is that zip urls will be gone soon enough.
